pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,603 @@
+		} else {
+			resp, err := delta.Recv()
+			if err != nil {
+				log.Error("Failed to receive message from stream", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+				continue
+			}
+			select {
+			case <-ctx.Done():
+				return
+			case c.deltaResponseQueue <- resp:
+			}
+		}
+	}
+}
+
+func (c *XDSClient) getAllResources(typeUrl string) (*xds.VersionedResources, error) {
+	return c.cache.GetResources(typeUrl, 0, "", nil)
+}
+
+// Observe adds resourceNames to watched resources of a given typeUrl.
+// It will be sent to a server asynchronyously.
+func (c *XDSClient) Observe(ctx context.Context, typeUrl string, resourceNames []string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.observeQueue <- &observeRequest{typeUrl: typeUrl, resourceNames: resourceNames}:
+	}
+	return nil
+}
+
+func (c *XDSClient) prepareRequest(obsReq *observeRequest) (*discoverypb.DiscoveryRequest, error) {
+	curr, err := c.getAllResources(obsReq.typeUrl)
+	if err != nil {
+		return nil, fmt.Errorf("get resources: %w", err)
+	}
+	reqResourceNames := make(map[string]bool, len(curr.ResourceNames)+len(obsReq.resourceNames))
+	for _, rn := range curr.ResourceNames {
+		reqResourceNames[rn] = true
+	}
+	for _, rn := range obsReq.resourceNames {
+		reqResourceNames[rn] = true
+	}
+
+	return &discoverypb.DiscoveryRequest{
+		Node:          c.node,
+		TypeUrl:       obsReq.typeUrl,
+		ResourceNames: slices.Collect(maps.Keys(reqResourceNames)),
+	}, nil
+}
+
+// loop will process responses from stream until Context ctx is done.
+// Errors are logged and processing continues.
+func (c *XDSClient) loop(ctx context.Context, errCh chan error,
+	stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient,
+	delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient,
+) {
+	defer close(errCh)
+	log := c.log.With(logFeatureKey, "loop")
+	backoff := backoff.Exponential{
+		Min:        c.opts.MinBackoff,
+		Max:        c.opts.MaxBackoff,
+		ResetAfter: c.opts.BackoffReset,
+		Jitter:     true,
+		Name:       "xds-client-loop",
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case obsReq, ok := <-c.observeQueue:
+			if !ok {
+				return
+			}
+			var err error
+			if c.opts.UseSOTW {
+				var req *discoverypb.DiscoveryRequest
+				req, err = c.prepareRequest(obsReq)
+				if err != nil {
+					log.Error("Failed to prepare request", "observe-request", obsReq, "error", err)
+					continue
+				}
+				log.Debug("Stream send", "req", req)
+				err = stream.Send(req)
+			} else {
+				req := &discoverypb.DeltaDiscoveryRequest{
+					Node:                   c.node,
+					TypeUrl:                obsReq.typeUrl,
+					ResourceNamesSubscribe: obsReq.resourceNames,
+				}
+				log.Debug("Delta send", "req", req)
+				err = delta.Send(req)
+			}
+			if err != nil {
+				log.Error("Stream/Delta send", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+			}
+		case resp, ok := <-c.responseQueue:
+			if !ok {
+				return
```

Comment:
should we log something here?

Reply:
No. This a case when we are stopping the client